### DISCUSSION TITLE
Opening CLIB4 Shared Objects and clib4.library

### DISCUSSION THREAD

**Discussion by 3246251196** [AmigaLabs/clib4#212] [Ideas]
### Shared Objects
We should agree on some sort of standard for opening _CLIB4_ _SOs_. Right now, our dynamic loader will look in the **PWD** or the **PWD/sobjs** directory in preference to **SOBJS:**. That would mean that we would have to distribute all programs built with _CLIB4_ with the _SOs_ in either **PWD** or **PWD/sobjs**.

I am assuming that it would be a bad idea to put _any_ _CLIB4_ _SOs_ into **SOBJS:** since that location should be kept for _NEWLIB_.

We could modify _elf.library_ to check in **SOBJS:clib4**, but we would need to get _elf.library_ to do that only when it knows that the binary was built with _CLIB4_.

We could implement _RPATH_ in _elf.library_. That would mean that we would need to remember to apply that mechanism when linking the program. Or, we could implicitly do this _always_ via changing _ADTOOLS_.

Any more options or ideas - other than the obvious and - my personal favourite - of linking statically, always!

### clib4.library
Whether linking statically or dynamically, this is relevant. Unlike _NEWLIB_, _CLIB4_ is changing rapidly in relative terms. Either *clib4.library* is already resident, or it exists in *LIBS:*. We could force a transient modification to *LIBS:* to add the *PWD* to *LIBS* so that _OpenLibrary_ would also check *PWD* in the case that a particular *clib4.library* is distributed.

**Comment by walkero-gr** [comment]
About so and the library files, I was under the impression that the system always checks first the PWD and then the sys: drive. Isn't that the case?

Also, what happens when an app loads clib4.library from libs: and then another app loads it from its folder? I believe that since it is in memory, the second app will use this one and will not load it again, unless there is a restriction on the version.

**Reply by smarkusg** [reply]
* library are always loaded first from ‘SYS’.
*so checks first PWD:LIbS ... PWD then SYS

**Reply by 3246251196** [reply]
Hey, @walkero-gr.

Right, for _SOs_ then it will look into the binary's location first. So, this is fine. If we are happy to do that, then fine.

As for _shared libraries_ - that is a different story. If it is not in memory already, it will check *LIBS:* unless you hack that assign to include an additional location, which is what I used to do for _adtools\_testing_. (**EDIT**: Not true since CLIB4 uses RomTags. There are a bunch of other paths it checks, starting with: Memory resident, if not, LIBS:, if not ...etc... if not, PWD:, if not PWD:Libs...)

This is just about a standard approach. If we are happy to just distribute _SOs_ with the program, then fine; but you better not forget to put it there, otherwise a small nuclear explosion may occur with a mess of the loader using _NEWLIB_ _SOs_ from *SOBJS:*

**Reply by smarkusg** [reply]
Current dynamic libraries (.so) are better to use static libraries in clib4 to avoid, as rjd324 wrote, the problem of loading libraries from NEWLIB

**Reply by afxgroup** [reply]
I don't think we can do a lot if Ramlib/Elf cannot be patched/changed. At moment clib4 .so and .library works like any other OS4 component. I don't know how you can change the behaviour without touching the OS

**Reply by 3246251196** [reply]
Exactly. So, would you be happy to always distribute SOs along with zip'd program, or should we definitely change our dynamic loader?

**Comment by 3246251196** [comment]
This is why I could not find the discussion, because it was on CLIB4. Fine. Anyway, we have made the decision to go ahead with infixing name/versions into the SO files. I will close this with this comment, but feel free to re-open if needed.

**Reply by migthymax** [reply]
[All those wondering where the discussion took place](https://github.com/AmigaLabs/adtools/issues/16)
